### ISSUE TITLE
Recognize "Godot Box2D" (Box2D) as an officially endorsed add-on for 2D physics

### ISSUE THREAD

**Issue by Ughuuu** [godotengine/godot-proposals#7415]
### Describe the project you are working on

A 2d cooking game with soft bodies. It has a high count of rigidbodies.

### Describe the problem or limitation you are having in your project

The current Physics 2D Engine has the current limitations:

- Is unstable at high count of rigidbodies.(unless you increase the iterations/simulations fps)
- Is unstable at low count of joints(joints just don't work that well unless you put really low restraints or increase the iterations/simulation fps).
- Lacks a lot of the newer 2d joints that would be nice to have.

Box2D is on the left. Godot Physics Engine in on the right.
![output (online-video-cutter com)](https://github.com/godotengine/godot-proposals/assets/2369380/0cff4721-2f4b-462a-bde4-6886715aa87e)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

[Box2D](https://github.com/erincatto/box2d) is a physics engine made by [Erin Catto](https://github.com/erincatto) and published under the [MIT license](https://en.wikipedia.org/wiki/MIT_license). It has been used in [Crayon Physics](https://en.wikipedia.org/wiki/Crayon_Physics) Deluxe, [Limbo](https://en.wikipedia.org/wiki/Limbo_(video_game)), etc.

Also, @erincatto is working on [box2c](https://github.com/erincatto/box2c), an improvement over box2d or a version 3 That will offer certain performance improvements.

The [Godot Box2D](https://github.com/godot-box2d/godot-box2d) is a Godot extension that integrates the box2d physics engine. It's not yet at feature parity, but it's getting close. It was started by @rburing and now implemented by me.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Similar to [Godot Jolt](https://github.com/godotengine/godot-proposals/issues/7308) proposal.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, it needs a whole godot extension to replace the current godot 2d physics.

### Is there a reason why this should be core and not an add-on in the asset library?

It's an improvement in everyway over the default godot 2d physics.

**Comment by reduz** [comment]
GodotPhysics2D has waaaaaay too many small hacks for usability and code that took an eternity to finetune, especially for character controllers.

It may not be able to push as many boxes, but its extremely predictable and intuitive. I am afraid achieving feature parity at the same level Jolt does is a monumental task for 2D. This is most likely what would be expected for any kind of official recognition.

**Comment by reduz** [comment]
On the positive side, I believe we had a good amount of unit tests for a lot of these functions, so that may help.

**Comment by Ughuuu** [comment]
> GodotPhysics2D has waaaaaay too many small hacks for usability and code that took an eternity to finetune, especially for character controllers.
> 
> It may not be able to push as many boxes, but its extremely predictable and intuitive. I am afraid achieving feature parity at the same level Jolt does is a monumental task for 2D. This is most likely what would be expected for any kind of official recognition.

Agree, I have noticed there are cases where the character controller does not stick right to the ground, or if it goes round between 4 walls it loses a small amount of distance. But I think/hope these can be fixed. Not saying it should be added right away, but maybe in the future, when it's stable. Opened this ahead to start a discussion about this.

**Comment by jams3223** [comment]
> Short update on this. Progress is going ok, implemented almost everything except multithreading, as box2d doesn't really have support for that right now.
> 
> Erin Catto, the maker of Box2D said once he finishes the new/updated version(Version 3) of box2d, he will also help with the tests to pass(or at least for the physics add-on to work correctly). Also, the new/updated version will have multithreading support.
> 
> Right now, from tests, I pass around 132, out of which 10 of them are very close to pass(or box2d doesn't offer an option for it, eg. sleep parameter is hardcoded in box2d, but in godot it has another value). 8 tests are still failing.

Wow, splendid work! Where is the repository for the Box2D v3 physics engine so I can monitor its progress?

**Comment by Ughuuu** [comment]
It's called [box2c](https://github.com/erincatto/box2c). If you are interested on its progress, you can join box2d discord channel and view the updates there from its creator. [Discord](https://discord.gg/NKYgCBP)

**Comment by EsdrasCaleb** [comment]
> GodotPhysics2D has waaaaaay too many small hacks for usability and code that took an eternity to finetune, especially for character controllers.
> 
> It may not be able to push as many boxes, but its extremely predictable and intuitive. I am afraid achieving feature parity at the same level Jolt does is a monumental task for 2D. This is most likely what would be expected for any kind of official recognition.

i don't get it. Do you will update the 2d physics small bugs or you want to test jolt to have the same optimizations?

**Comment by Norrox** [comment]
Love it so far! works great :D
![image](https://github.com/godotengine/godot-proposals/assets/13393087/0fcedafa-7ffe-4bfe-beb8-96883a195be9)

**Comment by Pizzaandy** [comment]
Update on this issue since it was mentioned in dev chat: I've been working on a [Box2D v3 plugin](https://github.com/Pizzaandy/godot-box2d-v3) that I think is ready for general review (including character movement!).

Given that Godot Physics 2D is based on Box2D 2.4 (it uses the same type of solver), I can see why it wouldn't seem necessary to endorse Box2D as an alternative - but Box2D v3 is an entirely new engine and a generational step up in terms of simulation speed and accuracy. Additionally, Erin has been very helpful in implementing features needed for Godot, including a new sensor system for Areas. 

I think there's a lot to gain from adopting Box2D - but regardless of what happens with 2D physics moving forward, I'm happy to continue maintaining Godot Box2D v3 :)
